### PR TITLE
accounts-db: throw error explicitly for next_account_offset calculation when overflows

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem
1. When we scanning append_vec, `next_account_offset` calculation is not safe, and it assumes that calculation never overflow. If it overflows, the app will crash. Fortunately, this is only an issue for tests and debug tools. For production validator code, we always sanitize the storage when we load. Therefore, we should (hopefully)never hit overflow for `new_account_offset` calculation. However, it would still be nice to throw the error explicitly when we overflow. 
1. We don't have unit test to cover `scan_index`. And the current test coverage doesn't cover corner cases. 
image

#### Summary of Changes
1. throw overflow error explicitly when calculating `next_account_offset`.
2. add test to cover `scan_index` with the corner cases, such incomplete account's data and garbage data that cause overflow. (now split into #2123)
3. improve `test_scan_pubkey` to cover mroe corner cases (now split into #2123).

Fixes #

### CODE REVIEW

File: accounts-db/src/append_vec.rs [hunk 1]
```diff
@@ -437,9 +437,14 @@ impl AppendVec {
                 self.flush().expect("flush must succeed");
                 // we are re-opening the file, so don't remove the file on disk when the old mmapped one is dropped
                 self.remove_file_on_drop.store(false, Ordering::Release);
-                AppendVec::new_from_file(self.path.clone(), self.len(), StorageAccess::File)
-                    .ok()
-                    .map(|(av, _size)| av)
+
```
Comment (HaoranYi) [inline accounts-db/src/append_vec.rs]: This is split into a separate PR: https://github.com/anza-xyz/agave/pull/2098

Comment (brooksprumo) [reply accounts-db/src/append_vec.rs]: Can you remove it from this PR then?

Comment (HaoranYi) [reply accounts-db/src/append_vec.rs]: done.

File: accounts-db/src/append_vec.rs [hunk 2]
```diff
@@ -840,17 +840,23 @@ impl AppendVec {
...
-    fn next_account_offset(start_offset: usize, stored_meta: &StoredMeta) -> AccountOffsets {
-        let stored_size_unaligned = STORE_META_OVERHEAD + stored_meta.data_len as usize;
+    ///
+    /// For safety, when `stored_meta.data_len` cause the calculation to overflow, Return None.
+    fn next_account_offset(
+        start_offset: usize,
+        stored_meta: &StoredMeta,
+    ) -> Option<AccountOffsets> {
+        let stored_size_unaligned =
+            STORE_META_OVERHEAD.checked_add(stored_meta.data_len as usize)?;
```
Comment (brooksprumo) [inline accounts-db/src/append_vec.rs]: I don't think we'll ever actually have an overflow here. Account data has a maximum of 10 MiB, and I think the sanitization checks when we open append vecs should catch a malformed append vec (hopefully?), right?

I'd be ok with doing a `checked_add().unwrap()` or `checked_add().expect()`, but I think we'd want to identify a bad append vec earlier, and then require/enforce than invariant here.

The *callers* of next_account_offset() can/should handle if the return value exceeds the file size.

File: accounts-db/src/append_vec.rs [hunk 3]
```diff
@@ -1938,16 +1950,121 @@ pub mod tests {
...
+        // Append a garbage AccountMeta to simulate next_offset overflow.
+        {
+            let mut f = OpenOptions::new()
+                .read(true)
+                .append(true)
+                .open(&temp_file.path)
+                .unwrap();
+            f.write_all(&[0xFF; SIZE]).unwrap();
+        }
+        assert_scan(total_stored_size + SIZE + 1);
```
Comment (brooksprumo) [inline accounts-db/src/append_vec.rs]: Also, I personally prefer unit tests to test a single thing. This way they can be smaller, so faster to run and easier to understand. I think these cases with bad/invalid append vecs/data could be their own tests.

File: accounts-db/src/append_vec.rs (hunk 2 above)
Comment (HaoranYi) [reply accounts-db/src/append_vec.rs]: sounds good. 
I updated the code to explicitly throw when we overflow. And also, add a comment on the fn about the overflow safety.

File: accounts-db/src/append_vec.rs (hunk 3 above)
Comment (HaoranYi) [reply accounts-db/src/append_vec.rs]: Yeah. Refactor the tests to share common code and separate the tests into multiple tests. Each tests one scenario.

File: accounts-db/src/append_vec.rs [hunk 4]
```diff
@@ -928,7 +932,10 @@ impl AppendVec {
                         Self::get_type(bytes_subset, next).unwrap();
                     let (_hash, next): (&AccountHash, _) =
                         Self::get_type(bytes_subset, next).unwrap();
-                    let stored_size_aligned = u64_align!(next + (meta.data_len as usize));
+                    let stored_size_aligned = next
```
Comment (HaoranYi) [inline accounts-db/src/append_vec.rs]: Make the next element offset calculation to throw the matched error string. This makes test assert panic string easier.

File: accounts-db/src/append_vec.rs (hunk 2 above)
Comment (brooksprumo) [reply accounts-db/src/append_vec.rs]: I like it.

I think this change can be its own PR. Or in other words, this PR currently feels like two sets of changes. One is this checked_add() here, and the other is the tests. Can they be in separate PRs?

For the `expect` message here, can it use the "expect as precondition" style, as recommended here: https://doc.rust-lang.org/std/error/index.html#common-message-styles ?

Something like "stored size cannot overflow"?

File: accounts-db/src/append_vec.rs (hunk 3 above)
Comment (brooksprumo) [reply accounts-db/src/append_vec.rs]: Nice. I'm adding a test for scan_index() over here too: https://github.com/anza-xyz/agave/pull/2121

File: accounts-db/src/append_vec.rs (hunk 2 above)
Comment (HaoranYi) [reply accounts-db/src/append_vec.rs]: yeah. updated.

Comment (HaoranYi) [reply accounts-db/src/append_vec.rs]: Now the tests are split into #2123.

File: accounts-db/src/append_vec.rs [hunk 5]
```diff
@@ -928,7 +932,10 @@ impl AppendVec {
...
+                        .checked_add(meta.data_len as usize)
+                        .expect("stored size cannot overflow");
+                    let stored_size_aligned = u64_align!(stored_size_aligned);
```
Comment (brooksprumo) [inline accounts-db/src/append_vec.rs]: I'm updating `scan_index()` to unify the checks in https://github.com/anza-xyz/agave/pull/2122. Wdyt about reverting this for now? With #2122, it'll use `next_account_offset()` and will inherit the checked math.

Comment (HaoranYi) [reply accounts-db/src/append_vec.rs]: Ah. I see. 
Yeah. Reverted!

Comment (HaoranYi) [reply accounts-db/src/append_vec.rs]: Oops. Committed the revert in the other branch. 
Now cherry-picked it here too.

**Review by brooksprumo — APPROVED** [review]
:shipit:
